docs/posts/pydata-performance/index.qmd

```diff
@@ -0,0 +1,696 @@
+a bunch of data we're not going to use. We can figure these columns out by
+manually inspecting the ibis code above:
+
+```{python}
+columns = ["path", "uploaded_on", "project_name"]
+
+%time df = pd.read_parquet(smallest_file, columns=columns)
+df
+```
+
+Sweet, read times improved!
+
+Let's looks at the memory usage of the `DataFrame`.
+
+```{python}
+df.memory_usage().sum() / (1 << 20)  # MiB
+```
+
+I still have plenty of space to do my analysis, nice!
+
+First, filter the data:
+
+```{python}
+%%time
+df = df[
+    (
+        df.path.str.contains(r"\.(?:asm|c|cc|cpp|cxx|h|hpp|rs|[Ff][0-9]{0,2}(?:or)?|go)$")
+        & ~df.path.str.contains(r"(?:^|/)test(?:|s|ing)|/site-packages/")  # <1>
+    )
+]
+df
+```
+
+1. I altered the original query here to avoid creating an unnecessary
+   intermediate `Series` object.
+
+We've blown **way** past our ~30s Ibis + DuckDB latency budget.
+
+Let's keep going!
+
+Next, group by and aggregate:
+
+```{python}
+#| error: false
+%%time
+df = df.groupby(
+    [
+        df.uploaded_on.dt.floor("M"),
+        df.path.str.extract(r"\.([a-z0-9]+)$", 0),
+    ]
+)
+df
+```
+
+Here we hit the first API issue going back to an [old pandas
+issue](https://github.com/pandas-dev/pandas/issues/15303): we can't truncate
+a timestamp column to month frequency.
+
+Let's try the solution recommended in that issue.
+
+```{python}
+#| error: false
+%%time
+df = df.groupby(
+    [
+        df.uploaded_on.dt.to_period("M").dt.to_timestamp(),
+        df.path.str.extract(r"\.([a-z0-9]+)$", 0)
+    ]
+)
+df
+```
+
+That seems to have gotten past the `floor` issue but now I get an error about
+the shape of the grouping keys. I suspect this is because `extract` returns
+a `DataFrame`. Such is life.
+
+Let's try a different strategy: compute the columns in an `assign` call and
+then do the group by using string column names:
+
+```{python}
+%%time
+df = (
+    df.assign(
+        month=lambda t: t.uploaded_on.dt.to_period("M").dt.to_timestamp(), # <1>
+        ext=lambda t: t.path.str.extract(r"\.([a-z0-9]+)$", 0), # <1>
+    )
+    .groupby(["month", "ext"])
+    .agg({"project_name": lambda s: list(set(s))})
+    .rename(columns={"project_name": "projects"})
+)
+df
+```
+
+1. Later we'll see that Dask doesn't support lambda functions in `assign` and
+   instead we'll create an intermediate `DataFrame` for the `assign` call.
+
+Sort the values, add a new column and do the final aggregation:
+
+```{python}
+%%time
+df = (
+    df.reset_index()  # <1>
+    .sort_values("month", ascending=False)
+    .assign(
+        ext=lambda t: t.ext.str.replace(r"cxx|cpp|cc|c|hpp|h", "C/C++", regex=True)
+        .str.replace("^f.*$", "Fortran", regex=True)
+        .str.replace("rs", "Rust")
+        .str.replace("go", "Go")
+        .str.replace("asm", "Assembly")
+        .replace("", None)
+    )
+    .groupby(["month", "ext"])
+    .agg({"projects": lambda s: len(set(sum(s, [])))})
```

Comment:
What is this line doing?

Reply:
Why are you using lambdas before you switch to more efficient things?

Reply:
Because that replicates the original query.

Reply:
Same as https://github.com/ibis-project/ibis/pull/7624/files#r1406490006: this is intentional (and later I show that this is not really the best way to accomplish our goal anyway)